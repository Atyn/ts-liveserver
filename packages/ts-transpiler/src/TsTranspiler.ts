import Fs from 'fs'
import Path from 'path'
import TypeScript from 'typescript'
import CompilerOptions from './CompilerOptions'
import ResolveTransformer from './transformers/ResolveTransformer'
import CommonJsTransformer from './transformers/CommonJsTransformer'
import NodeEnvTransformer from './transformers/NodeEnvTransformer'
import CodeOptimizerTransformer from './transformers/CodeOptimizerTransformer'
import InternalDependencyReducer from './transformers/InternalDependencyReducer'
import DependencyResolver from './DependencyResolver'
import LanguageService from './LanguageService'

const RESOLVE_EXTENSIONS = ['.js', '.ts', '.tsx', '.jsx', '.json', '.mjs']

type Options = {
	compilerOptions?: TypeScript.CompilerOptions
	resolveAlias?: Record<string, string>
}
export default class TsTranspiler {
	private languageService: LanguageService
	private compilerOptions: TypeScript.CompilerOptions = CompilerOptions
	constructor(options?: Options) {
		const dependencyResolver = new DependencyResolver(options?.resolveAlias)
		Object.assign(this.compilerOptions, options?.compilerOptions, {
			rootDir: '../..',
		})
		this.languageService = new LanguageService({
			dependencyResolver: dependencyResolver,
			compilerOptions: this.compilerOptions,
			transformers: {
				before: [
					(context) => new NodeEnvTransformer(context),
					(context) => new CodeOptimizerTransformer(context),
				],
				after: [
					(context) => new CommonJsTransformer(context),
					(context) => new ResolveTransformer(context, dependencyResolver),
				],
			},
		})
	}
	public async transformFile(
		fileName: string,
	): Promise<TypeScript.TranspileOutput> {
		if (Path.extname(fileName) === '.json') {
			const fileContent =
				'export default ' + (await Fs.promises.readFile(fileName))
			return {
				outputText: fileContent,
			}
		}
		return await this.languageService.transformFile(fileName)
	}
	public async resolveFilePath(fileName: string): Promise<string> {
		if (await this.fileExists(Path.resolve(fileName))) {
			return Path.resolve(fileName)
		}
		const parsedFileName = Path.parse(fileName)
		for (const extension of RESOLVE_EXTENSIONS) {
			if (
				await this.fileExists(
					Path.resolve(parsedFileName.dir, parsedFileName.name + extension),
				)
			) {
				return Path.resolve(parsedFileName.dir, parsedFileName.name + extension)
			}
		}
		throw new Error('Could not find file' + fileName)
	}
	private async fileExists(path: string) {
		try {
			await Fs.promises.stat(path)
		} catch (error) {
			return false
		}
		return true
	}
}
