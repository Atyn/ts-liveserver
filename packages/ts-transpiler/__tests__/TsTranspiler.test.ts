import TsTranspiler from '../src/TsTranspiler'
import Path from 'path'
import Fs from 'fs'

describe('TsTranspiler', () => {
	it('Transpile a module', async () => {
		const inputFileName = './ModuleA.ts'
		const transpiler = new TsTranspiler()
		console.log(Path.resolve(__dirname, 'inputData', inputFileName))
		const results = await transpiler.transformFile(
			Path.resolve(__dirname, 'inputData', inputFileName),
		)
		expect(results).toBe(
			Fs.readFileSync(
				Path.resolve(__dirname, 'outputData', 'ModuleA.js'),
			).toString(),
		)
	})
})
