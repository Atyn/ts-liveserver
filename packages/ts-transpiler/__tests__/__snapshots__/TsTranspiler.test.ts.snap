// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TsTranspiler Transpile a module 1`] = `
"/* eslint-disable */
import * as React from \\"../../../../node_modules/react/index.js\\";
import ModuleB from \\"./ModuleB.js\\";
const component = React.createElement(\\"div\\", null, \\"Hello\\");
console.log(component);
const ModuleA = ModuleB;
export default ModuleA;
"
`;
